rfcs/0000-custom-underlying-services.md

```diff
@@ -0,0 +1,163 @@
+
+Referencing a custom resource provides good level of abstraction and decoupling. The kuadrant control plane must obtain all required information from the targeted custom resources to implement the integration of the services in the data plane. For instance, the Limitador CR provides useful information in the `status` section. First and foremost, if the service is available (condition type `Ready`).
+
+```yaml=
+apiVersion: limitador.kuadrant.io/v1alpha1
+kind: Limitador
+metadata:
+  name: limitador
+  namespace: kuadrant-system
+spec: {}
+status:
+  conditions:
+  - lastTransitionTime: "2023-05-02T13:23:45Z"
+    message: Limitador is ready
+    reason: Ready
+    status: "True"
+    type: Ready
+  observedGeneration: 1
+  service:
+    host: limitador-limitador.kuadrant-system.svc.cluster.local
+    ports:
+      grpc: 8081
+      http: 8080
+```
+
+**Only one reference**
+
+From a kuadrant CR you can only **reference one**, and only one, limitador and authorino instance. This is a design decision taken for the current needs and powered by its simplicity. In the future, if there are use cases for that, one kuadrant instance could manage multiple, let's say Limitador, instances and assign those instances to the managed gateways based on some rules. Something like, this gateway A gets assigned Limitador A and this gateway B gets assigned Limitador B for whatever reason. Currently, this scenario is out of scope.
+
+**References are optional**
+
+The references are **optional**. In the absence of a reference, the Kuadrant operator will deploy and operate a default deployment (yet to be defined) to support the kuadrant policies.
+
+---
+Explain the proposal as if it was implemented and you were teaching it to Kuadrant user. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how a user should *think* about the feature, and how it would impact the way they already use Kuadrant. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing and new Kuadrant users.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+TODO
+
+---
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- How error would be reported to the users.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+TODO
```

Comment:
One potential drawback is that we do not want users to change some settings. From the top of my head, I can think of Authorino `spec.clusterWide` and `authConfigLabelSelectors`, which could mean that the user could easily end up with AuthPolicy that will be reconciled by Kuadrant but won't be picked up in Authorino.

Reply:
Kuadrant should reconcile those changes in the external services, right? More specifically, when the Authorino CR is (externally) updated and the label selectors are updated, the AuthConfig CR's managed by kuadrant should reconcile to match.